src/group_theory/semidirect_product.lean

```diff
@@ -0,0 +1,134 @@
+  group of `N`. It the product of sets with the group operation
+  `⟨n₁, g₁⟩ * ⟨n₂, g₂⟩ = ⟨n₁ * φ g₁ n₂, g₁ * g₂⟩` -/
+@[ext, derive decidable_eq]
+structure semidirect_product (φ : G →* mul_aut N) :=
+(left : N) (right : G)
+
+attribute [pp_using_anonymous_constructor] semidirect_product
+
+notation N` ⋊[`:35 φ:35`] `:0 G :35 := semidirect_product N G φ
+
+namespace semidirect_product
+
+variables {N G} {φ : G →* mul_aut N}
+
+private def one_aux : N ⋊[φ] G := ⟨1, 1⟩
+private def mul_aux (a b : N ⋊[φ] G) : N ⋊[φ] G := ⟨a.1 * φ a.2 b.1, a.right * b.right⟩
+private def inv_aux (a : N ⋊[φ] G) : N ⋊[φ] G := let i := a.2⁻¹ in ⟨φ i a.1⁻¹, i⟩
+private lemma mul_assoc_aux (a b c : N ⋊[φ] G) : mul_aux (mul_aux a b) c = mul_aux a (mul_aux b c) :=
+by simp [mul_aux, mul_assoc, mul_equiv.map_mul]
+private lemma mul_one_aux (a : N ⋊[φ] G) : mul_aux a one_aux = a :=
+by cases a; simp [mul_aux, one_aux]
+private lemma one_mul_aux (a : N ⋊[φ] G) : mul_aux one_aux a = a :=
+by cases a; simp [mul_aux, one_aux]
+private lemma mul_left_inv_aux (a : N ⋊[φ] G) : mul_aux (inv_aux a) a = one_aux :=
+by simp only [mul_aux, inv_aux, one_aux, ← mul_equiv.map_mul, mul_left_inv]; simp
+
+instance : group (N ⋊[φ] G) :=
+{ one := one_aux,
+  inv := inv_aux,
+  mul := mul_aux,
+  mul_assoc := mul_assoc_aux,
+  one_mul := one_mul_aux,
+  mul_one := mul_one_aux,
+  mul_left_inv := mul_left_inv_aux }
+
+instance : inhabited (N ⋊[φ] G) := ⟨1⟩
+
+@[simp] lemma one_left : (1 : N ⋊[φ] G).left = 1 := rfl
+@[simp] lemma one_right : (1 : N ⋊[φ] G).right = 1 := rfl
+@[simp] lemma inv_left (a : N ⋊[φ] G) : (a⁻¹).left = φ a.right⁻¹ a.left⁻¹ := rfl
+@[simp] lemma inv_right (a : N ⋊[φ] G) : (a⁻¹).right = a.right⁻¹ := rfl
+@[simp] lemma mul_left (a b : N ⋊[φ] G) : (a * b).left = a.left * φ a.right b.left := rfl
+@[simp] lemma mul_right (a b : N ⋊[φ] G) : (a * b).right = a.right * b.right := rfl
+
+/-- The canonical map `N →* N ⋊[φ] G` sending `n` to `⟨n, 1⟩` -/
+def inl : N →* N ⋊[φ] G :=
+{ to_fun := λ n, ⟨n, 1⟩,
+  map_one' := rfl,
+  map_mul' := by intros; ext; simp }
+
+@[simp] lemma left_inl (n : N) : (inl n : N ⋊[φ] G).left = n := rfl
+@[simp] lemma right_inl (n : N) : (inl n : N ⋊[φ] G).right = 1 := rfl
+
+/-- The canonical map `G →* N ⋊[φ] G` sending `g` to `⟨1, g⟩` -/
+def inr : G →* N ⋊[φ] G :=
+{ to_fun := λ g, ⟨1, g⟩,
+  map_one' := rfl,
+  map_mul' := by intros; ext; simp }
+
+@[simp] lemma left_inr (g : G) : (inr g : N ⋊[φ] G).left = 1 := rfl
+@[simp] lemma right_inr (g : G) : (inr g : N ⋊[φ] G).right = g := rfl
+
+lemma inl_aut (g : G) (n : N) : (inl (φ g n) : N ⋊[φ] G) = inr g * inl n * inr g⁻¹ :=
+by ext; simp
+
+lemma inl_left_mul_inr_right (x : N ⋊[φ] G) : inl x.left * inr x.right = x :=
+by ext; simp
+
+/-- The canonical projection map `N ⋊[φ] G →* G`, as a group hom. -/
+def right_hom : N ⋊[φ] G →* G :=
+{ to_fun := semidirect_product.right,
+  map_one' := rfl,
+  map_mul' := λ _ _, rfl }
+
```

Comment:
Please add proofs that `inl` and `inr` are injective, and that `right_hom` is surjective.
Maybe also show the short exact sequence `1 → N -inl→ N ⋊[φ] G -right_hom→ G → 1` in some form...